### PR TITLE
Add TiffImages for TIFFs

### LINKED ISSUE #18: Add TiffImages for TIFFs
We should add https://github.com/tlnagy/TiffImages.jl so that we can support TIFFs

@tlnagy is there anything preventing this on the TiffImages side?

### PR DESCRIPTION
**Pull request by IanButterworth**
@tlnagy this is a quick attempt to start this. I'd appreciate your help getting this working

The other packages provide `save` functions to go right from native julia image arrays to files. I don't believe TiffImages does? Perhaps that could be added and used here, instead of the way I constructed the `save` methods?

When finished:
closes #18

### CODE REVIEW

File: src/ImageIO.jl [hunk 1]
```diff
@@ -69,4 +66,35 @@ for NETPBMFORMAT in (:PBMBinary, :PGMBinary, :PPMBinary, :PBMText, :PGMText, :PP
...
+end
+
+function save(s::Stream{DataFormat{:TIFF}}, image::S; permute_horizontal=false, mapi=identity) where {T, S<:Union{AbstractMatrix, AbstractArray{T,3}}}
+    imgout = map(mapi, image)
+    tiff_imgout = Base.invokelatest(checked_import(:TiffImages).DenseTaggedImage, imgout)
+    if permute_horizontal
+        perm = ndims(tiff_imgout) == 2 ? (2, 1) : ndims(tiff_imgout) == 3 ? (2, 1, 3) : error("$(ndims(tiff_imgout)) dims array is not supported")
+        open(stream(s), "w") do io ## ???????
+            Base.invokelatest(checked_import(:TiffImages).write, io, PermutedDimsArray(tiff_imgout, perm))
+        end
```
Comment (IanButterworth) [inline src/ImageIO.jl]: I'm not expecting this to work, but wasn't sure how to save into a stream

**Comment by tlnagy** [comment]
> The other packages provide `save` functions to go right from native julia image arrays to files. I don't believe TiffImages does? Perhaps that could be added and used here, instead of the way I constructed the `save` methods?

I agree that adding a `save` function to `TiffImages` is the way to fix this. The current API for saving is unnecessarily clunky.

**Comment by tlnagy** [comment]
I whipped together a PR over at https://github.com/tlnagy/TiffImages.jl/pull/29 that should simplify your PR here.

**Comment by tlnagy** [comment]
The new version of TiffImages is off to the presses. If you set the compat bounds starting at `v0.2.2`, you should be able to use the much simpler `save` API.

File: README.md [hunk 2]
```diff
@@ -6,8 +6,9 @@ FileIO.jl integration for image files
 [![Codecov](https://codecov.io/gh/JuliaIO/ImageIO.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/JuliaIO/ImageIO.jl)
 
 Currently provides:
-- PNGFiles.jl for Portable Network Graphics via libpng - ([Benchmark vs. ImageMagick & QuartzImageIO](https://github.com/JuliaIO/PNGFiles.jl/issues/1#issuecomment-586749654))
-- Netpbm.jl for Portable Bitmap formats (in pure Julia)
+- [PNGFiles.jl](https://github.com/JuliaIO/PNGFiles.jl) for Portable Network Graphics via libpng - ([Benchmark vs. ImageMagick & QuartzImageIO](https://github.com/JuliaIO/PNGFiles.jl/issues/1#issuecomment-586749654))
+- [Netpbm.jl](https://github.com/JuliaIO/Netpbm.jl) for Portable Bitmap formats (in pure Julia)
+- [TiffImages.jl](https://github.com/tlnagy/TiffImages.jl) for TIFFs
```
Comment (tlnagy) [inline README.md]: Worth adding that TiffImages is also in pure Julia?

**Comment by IanButterworth** [comment]
@tlnagy I think we're there! What do you think?

I added a TODO as a suggestion to add `UInt8` and `N0f8` eltype support in TiffImages, mainly because PNGFiles supports it.
It's non-critical though given they can always be wrapped as `Gray{N0f8}`

**Comment by tlnagy** [comment]
Looks good to me. There's a good case for automatically wrapping with Gray to match the other packages. I might get around to it at some point.

**Comment by IanButterworth** [comment]
Great. I'll get a release out and prepare a FileIO PR
Thanks!

**Pull request merged**